### PR TITLE
#830. enrichment tag hashes

### PR DESCRIPTION
**Pull request by madryk**

### CODE REVIEW

File: saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import pl.edu.icm.saos.enrichment.hash.JudgmentEnrichmentTags;
+import pl.edu.icm.saos.enrichment.hash.MarkProcessedJobExecutionListener;
+import pl.edu.icm.saos.enrichment.hash.UpdateEnrichmentHashProcessor;
+import pl.edu.icm.saos.enrichment.hash.UpdateEnrichmentHashReader;
+import pl.edu.icm.saos.enrichment.hash.UpdateEnrichmentHashWriter;
+import pl.edu.icm.saos.persistence.enrichment.model.JudgmentEnrichmentHash;
+
+@Configuration
+@ComponentScan
+public class TagPostUploadJobConfiguration {
```
Comment (lukdumi) [inline saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java]: No description.

File: saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    @Autowired
+    private UpdateEnrichmentHashProcessor updateEnrichmentHashProcessor;
+    
+    @Autowired
+    private UpdateEnrichmentHashWriter updateEnrichmentHashWriter;
+    
+    @Autowired
+    private MarkProcessedJobExecutionListener markProcessedJobExecutionListener;
+    
+
```
Comment (lukdumi) [inline saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java]: no logic section

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    @Autowired
+    private EntityManager entityManager;
+    
+    @Autowired
+    private EnrichmentTagRepository enrichmentTagRepository;
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    public List<JudgmentEnrichmentTags> fetchEnrichmentTagsForJudgments(List<Long> judgmentIds) {
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: no method comment

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import pl.edu.icm.saos.persistence.enrichment.model.EnrichmentTag;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+/**
+ * @author madryk
+ */
+@Service
+public class JudgmentEnrichmentTagsFetcher {
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: no class comment

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    private EnrichmentTagRepository enrichmentTagRepository;
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    public List<JudgmentEnrichmentTags> fetchEnrichmentTagsForJudgments(List<Long> judgmentIds) {
+        
+        List<EnrichmentTag> enrichmentTags = enrichmentTagRepository.findAllByJudgmentIds(judgmentIds);
+        
+        List<Object[]> enrichmentTagsWithReference = fetchJudgmentsReferencingTags(judgmentIds);
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: referencingTags?

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        collectJudgmentsEnrichmentTags(judgmentsEnrichmentTagsMap, enrichmentTags);
+        collectJudgmentsReferencingEnrichmentTags(judgmentsEnrichmentTagsMap, enrichmentTagsWithReference);
+        
+        return Lists.newLinkedList(judgmentsEnrichmentTagsMap.values());
+    }
+    
+    
+    //------------------------ PRIVATE --------------------------
+    
+    private List<Object[]> fetchJudgmentsReferencingTags(List<Long> judgmentIds) {
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: Why doesn't it return map?

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsHashCalculator.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    private final static String TAG_SEPARATOR = "::";
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    /**
+     * Returns hash of enrichment tags for given judgment based on md5 algorithm.
+     * If judgment has no enrichment tags then {@literal null} is returned.
+     */
+    public String calculateHash(JudgmentEnrichmentTags judgmentIdWithTags) {
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsHashCalculator.java]: judgmentIdWithTags -> judgmentEnrichmentTags?

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/MarkProcessedJobExecutionListener.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import pl.edu.icm.saos.persistence.enrichment.JudgmentEnrichmentHashRepository;
+
+/**
+ * @author madryk
+ */
+@Service
+public class MarkProcessedJobExecutionListener implements JobExecutionListener {
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/MarkProcessedJobExecutionListener.java]: EnrichmentHashProcessedFlagMarker?

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import javax.persistence.SequenceGenerator;
+import javax.persistence.Table;
+import javax.persistence.UniqueConstraint;
+
+import org.apache.commons.lang3.StringUtils;
+
+import pl.edu.icm.saos.persistence.common.DataObject;
+
+/**
+ * Entity containing hashes enrichment tags related to judgment.
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ...hashes generated from enrichment tags ...

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+ * Hashes are used to determine if enrichment tags has changed
+ * since last processing of it after their upload.
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ... to determine if the last upload of them has changed them.

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "seq_judgment_enrichment_hash")
+    @Override
+    public long getId() {
+        return id;
+    }
+
+
+    /**
+     * Returns id of judgment
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: link to judgment.getId()

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+        return oldHash;
+    }
+
+    /**
+     * Returns hash of judgment enrichment tags.
+     * If {@link #isProcessed()} is false then it contains
+     * newly calculated hash for current enrichment tags that are in database,
+     * but was not processed yet.
+     * If {@link #isProcessed()} is true then it contains
+     * hash for current enrichment tags and their processing have ended.
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ... tags of which processing has finished.

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public String getOldHash() {
+        return oldHash;
+    }
+
+    /**
+     * Returns hash of judgment enrichment tags.
+     * If {@link #isProcessed()} is false then it contains
+     * newly calculated hash for current enrichment tags that are in database,
+     * but was not processed yet.
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: but have not been processed yet

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public long getJudgmentId() {
+        return judgmentId;
+    }
+
+    /**
+     * Returns old hash of judgment enrichment tags.
+     * If {@link #isProcessed()} is false then it contains hash
+     * for last processed enrichment tags (new enrichment tags are
+     * in database, but was not processed yet)
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: but have not been processed yet.

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashReader.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        
+    }
+    
+    
+    //------------------------ PRIVATE --------------------------
+    
+    private void fetchPageOfJudgmentsEnrichmentTags() {
+        List<Long> judgmentIdsPage = Lists.newLinkedList();
+        
+        while(!judgmentIds.isEmpty() && judgmentIdsPage.size() < judgmentsEnrichmentTagsPageSize) {
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashReader.java]: judgmentIdsPage.size() < judgmentsEnrichmentTagsPageSize isn't it always true?

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    public String getHash() {
+        return hash;
+    }
+
+    /**
+     * Flag telling if judgment was processed with
+     * last known enrichment tags.
+     */
+    public boolean isProcessed() {
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: Returns true if the current judgment enrichment tags (for which #getHash has been generated) have been processed.

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+     * Returns id of judgment
+     */
+    public long getJudgmentId() {
+        return judgmentId;
+    }
+
+    /**
+     * Returns old hash of judgment enrichment tags.
+     * If {@link #isProcessed()} is false then it contains hash
+     * for last processed enrichment tags (new enrichment tags are
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ... for the last...

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+     * If {@link #isProcessed()} is true then it contains hash
+     * that is already outdated (new enrichment tags are in database
+     * and their processing have ended).
+     */
+    public String getOldHash() {
+        return oldHash;
+    }
+
+    /**
+     * Returns hash of judgment enrichment tags.
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ... the current judgment enrichment tags.

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * Returns id of judgment
+     */
+    public long getJudgmentId() {
+        return judgmentId;
+    }
+
+    /**
+     * Returns old hash of judgment enrichment tags.
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ...hash for the last but one judgment enrichment tags (not the current ones).

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    
+    private String hash;
+    
+    private boolean processed;
+
+    
+    //------------------------ LOGIC --------------------------
+    
+    /**
+     * Updates hash value. Marks judgment enrichment tags as not processed if
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: Updates the hash value (see: #getHash())...

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+        
+        
+        JudgmentEnrichmentHash enrichmentHash = judgmentEnrichmentHashRepository.findByJudgmentId(judgmentEnrichmentTags.getJudgmentId());
+        
+        if (enrichmentHash == null) {
+            enrichmentHash = new JudgmentEnrichmentHash();
+            enrichmentHash.setJudgmentId(judgmentEnrichmentTags.getJudgmentId());
+        }
+        
+        enrichmentHash.fillNewHash(hash);
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java]: doesn't compile, shouldn't it be updateHash?

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+        
+        List<EnrichmentTag> enrichmentTags = enrichmentTagRepository.findAllByJudgmentIds(judgmentIds);
+        
+        List<Object[]> enrichmentTagsWithReference = fetchJudgmentsReferencingTags(judgmentIds);
+        
+        Map<Long, JudgmentEnrichmentTags> judgmentsEnrichmentTagsMap = Maps.newHashMapWithExpectedSize(judgmentIds.size());
+        
+        initializeJudgmentsEnrichmentTagsMap(judgmentsEnrichmentTagsMap, judgmentIds);
+        collectJudgmentsEnrichmentTags(judgmentsEnrichmentTagsMap, enrichmentTags);
+        collectJudgmentsReferencingEnrichmentTags(judgmentsEnrichmentTagsMap, enrichmentTagsWithReference);
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: I would get rid of this map here.

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    
+    private JudgmentEnrichmentHashRepository judgmentEnrichmentHashRepository;
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    @Override
+    public JudgmentEnrichmentHash process(JudgmentEnrichmentTags judgmentEnrichmentTags) throws Exception {
+        
+        String hash = judgmentEnrichmentTagsHashCalculator.calculateHash(judgmentEnrichmentTags);
```
Comment (lukdumi) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java]: I'd move it just before the enrichmentHash.updateHash (to line 35).

**Comment by lukdumi** [comment]
JudgmentEnrichmentHash.updateHash method is very important and should be unit tested.

File: saos-enrichment/src/test/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessorTest.java [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+        
+        // assert
+        assertEquals(2L, retJudgmentEnrichmentHash.getJudgmentId());
+        assertEquals("hash", retJudgmentEnrichmentHash.getOldHash());
+        assertEquals("hash", retJudgmentEnrichmentHash.getHash());
+        assertTrue(retJudgmentEnrichmentHash.isProcessed());
+    }
+    
+    @Test
+    public void process_PROCESSED_WITH_CHANGE() throws Exception {
```
Comment (lukdumi) [inline saos-enrichment/src/test/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessorTest.java]: I suggest testing the change of hash in a separate unit test (JudgmentEnrichmentHashTest). Here I would only check if updateHash with a proper arg is invoked.

Comment (madryk) [inline saos-enrichment/src/test/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessorTest.java]: :+1:

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java (hunk 23 above)
Comment (madryk) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java]: :+1:

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java (hunk 21 above)
Comment (madryk) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/UpdateEnrichmentHashProcessor.java]: Yes, it should.

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsHashCalculator.java (hunk 7 above)
Comment (madryk) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsHashCalculator.java]: :+1:

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 20 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: :+1:

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 9 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: :+1:

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 10 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: :+1:

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 11 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ok

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 16 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ok

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 18 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ok

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java (hunk 19 above)
Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: ok

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java (hunk 4 above)
Comment (madryk) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: ok, I will add it

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java (hunk 3 above)
Comment (madryk) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/JudgmentEnrichmentTagsFetcher.java]: :+1:

File: saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/MarkProcessedJobExecutionListener.java (hunk 8 above)
Comment (madryk) [inline saos-enrichment/src/main/java/pl/edu/icm/saos/enrichment/hash/MarkProcessedJobExecutionListener.java]: ok

File: saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java (hunk 1 above)
Comment (madryk) [inline saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java]: :+1:

File: saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java (hunk 2 above)
Comment (madryk) [inline saos-batch/src/main/java/pl/edu/icm/saos/batch/core/TagPostUploadJobConfiguration.java]: :+1:

File: saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    public void updateHash(String newHash) {
+        if (processed) {
+            this.oldHash = this.hash;
+            this.hash = newHash;
+            if (!StringUtils.equals(this.oldHash, this.hash)) {
+                processed = false;
+            }
+        } else {
+            this.hash = newHash;
```
Comment (lukdumi) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: I wonder if the processed flag shouldn't be set here to true if this.hash == this.oldHash.

Comment (madryk) [inline saos-persistence/src/main/java/pl/edu/icm/saos/persistence/enrichment/model/JudgmentEnrichmentHash.java]: Yes, it would be correct.

**Pull request closed without merging**